### ISSUE TITLE
How to release CUDA memory cached by DALI?

### ISSUE THREAD

**Issue by cyyever** [NVIDIA/DALI#3396]
I am implementing a training system and I have found  that some CUDA memory is stilled cached by DALI after I have deleted all DALI objects. Since pytorch has its own allocators, those memory can be reused. I am not sure how to check the cached memory size and release them. Can someone tell me how to do?

**Comment by JanuszL** [comment]
Hi @cyyever,

If you remove all DALI pipeline and framework iterator the memory should be released.
Could you please share a minimal reproduction code that would show the problem you experience?
@mzient has anything changed with recent memory management rework?

**Comment by cyyever** [comment]
@mzient 
@JanuszL 
In my case, I fork a process for each worker to simulate distributed training. Since GPU resource is limited, a worker will release CUDA memory after training to give some other worker a chance to run. I have noticed the pool has a free_all method but is not exposed to Python currently.

**Comment by mzient** [comment]
`free_all` is not safe - in fact, it should be private. There's no function that frees unused blocks.

**Comment by cyyever** [comment]
@mzient 
I think there must be some form of memory leak. I have change the code to release and recreate dataset iterators repeatedly in a single thread, then I have observed CUDA memory usage increases slowly. This occurred after commit d3de6899c093393bbcc26e24d782fa06bcae1291 (I had updated and rebuild the source code.).

**Comment by cyyever** [comment]
@JanuszL 
 Here is a code snippet modified from 
```
DALI/docs/examples/use_cases/pytorch/resnet50/main.py
```
I changed the training loop to use a pipeline for 3 batches and then delete it and recreate. The validation code was deleted and others were left unchanged. I run this code on an A100 with 40 GB memory and it run out of memory in half a minute. 
The command line is
```
python3 main.py  ~/pytorch_dataset/ImageNet/ 
```
The nvidia-smi output is
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.57.02    Driver Version: 470.57.02    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-PCI...  On   | 00000000:21:00.0 Off |                    0 |
| N/A   35C    P0    35W / 250W |   9092MiB / 40536MiB |      3%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-PCI...  On   | 00000000:81:00.0 Off |                    0 |
| N/A   33C    P0    36W / 250W |      3MiB / 40536MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A   3292633      C   python3                          9097MiB |
+-----------------------------------------------------------------------------+
Sat Oct  9 14:57:58 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.57.02    Driver Version: 470.57.02    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-PCI...  On   | 00000000:21:00.0 Off |                    0 |
| N/A   38C    P0    54W / 250W |  17596MiB / 40536MiB |     34%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-PCI...  On   | 00000000:81:00.0 Off |                    0 |
| N/A   33C    P0    36W / 250W |      3MiB / 40536MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A   3292633      C   python3                         17589MiB |
+-----------------------------------------------------------------------------+
Sat Oct  9 14:58:04 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.57.02    Driver Version: 470.57.02    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-PCI...  On   | 00000000:21:00.0 Off |                    0 |
| N/A   38C    P0    54W / 250W |  22116MiB / 40536MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-PCI...  On   | 00000000:81:00.0 Off |                    0 |
| N/A   33C    P0    36W / 250W |      3MiB / 40536MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A   3292633      C   python3                         22109MiB |
+-----------------------------------------------------------------------------+
Sat Oct  9 14:58:09 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.57.02    Driver Version: 470.57.02    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-PCI...  On   | 00000000:21:00.0 Off |                    0 |
| N/A   39C    P0    54W / 250W |  26638MiB / 40536MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-PCI...  On   | 00000000:81:00.0 Off |                    0 |
| N/A   33C    P0    36W / 250W |      3MiB / 40536MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A   3292633      C   python3                         26631MiB |
+-----------------------------------------------------------------------------+
Sat Oct  9 14:58:14 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.57.02    Driver Version: 470.57.02    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-PCI...  On   | 00000000:21:00.0 Off |                    0 |
| N/A   39C    P0    54W / 250W |  31159MiB / 40536MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-PCI...  On   | 00000000:81:00.0 Off |                    0 |
| N/A   33C    P0    36W / 250W |      3MiB / 40536MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A   3292633      C   python3                         31151MiB |
+-----------------------------------------------------------------------------+
Sat Oct  9 14:58:20 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.57.02    Driver Version: 470.57.02    CUDA Version: 11.4     
[main.zip](https://github.com/NVIDIA/DALI/files/7315878/main.zip)
|
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-PCI...  On   | 00000000:21:00.0 Off |                    0 |
| N/A   39C    P0    54W / 250W |  35679MiB / 40536MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-PCI...  On   | 00000000:81:00.0 Off |                    0 |
| N/A   33C    P0    36W / 250W |      3MiB / 40536MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A   3292633      C   python3                         35671MiB |
+-----------------------------------------------------------------------------+
```

**Comment by JanuszL** [comment]
Hi @cyyever,

We tested on the latest [NVIDIA optimized Deep Learning Framework](https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel_21-09.html#rel_21-09):
```
NVIDIA Release 21.09 (build 27403213)
PyTorch Version 1.10.0a0+3fd9dcf
```
And most ToT (top of the tree) DALI build. If you see such a leak maybe it has something to do with how PyTorch deals with the memory allocated for Torch tensors that DALI produces.

**Comment by JanuszL** [comment]
Hi @cyyever,

DALI moved to CUDA 11.5 Update 1 and the nvJPEG leak has been fixed. Please check the latest nightly build that should include the https://github.com/NVIDIA/DALI/pull/3526 PR.
This fix should be also part of the DALI 1.9 release.

**Comment by JanuszL** [comment]
The issue should have been fixed in DALI 1.9.
